sdk/cosmos/azure-cosmos-tests/src/test/java/com/azure/cosmos/UserAgentSuffixTest.java

```diff
@@ -0,0 +1,79 @@
+/*
+ * Copyright (c) Microsoft Corporation. All rights reserved.
+ * Licensed under the MIT License.
+ *
+ */
+
+package com.azure.cosmos;
+
+import com.azure.cosmos.models.CosmosContainerResponse;
+import com.azure.cosmos.rx.TestSuiteBase;
+import com.fasterxml.jackson.core.json.JsonReadFeature;
+import com.fasterxml.jackson.databind.ObjectMapper;
+import org.testng.annotations.AfterClass;
+import org.testng.annotations.BeforeClass;
+import org.testng.annotations.Factory;
+import org.testng.annotations.Test;
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+public class UserAgentSuffixTest extends TestSuiteBase {
+    private static final ObjectMapper OBJECT_MAPPER = new ObjectMapper().configure(
+        JsonReadFeature.ALLOW_UNESCAPED_CONTROL_CHARS.mappedFeature(),
+        true
+    );
+
+    private CosmosClient client;
+    private String databaseName;
+    private String containerName;
+
+    @Factory(dataProvider = "clientBuildersWithDirectSessionIncludeComputeGateway")
+    public UserAgentSuffixTest(CosmosClientBuilder clientBuilder) {
+        super(clientBuilder.contentResponseOnWriteEnabled(true));
+    }
+
+    @BeforeClass(groups = {"fast"}, timeOut = SETUP_TIMEOUT)
+    public void before_CosmosItemTest() {
+        assertThat(this.client).isNull();
+        this.client = getClientBuilder().buildClient();
+        CosmosAsyncContainer asyncContainer = getSharedMultiPartitionCosmosContainer(this.client.asyncClient());
+        this.databaseName = asyncContainer.getDatabase().getId();
+        this.containerName = asyncContainer.getId();
+    }
+
+    @AfterClass(groups = {"fast"}, timeOut = SHUTDOWN_TIMEOUT, alwaysRun = true)
+    public void afterClass() {
+        assertThat(this.client).isNotNull();
+        this.client.close();
+    }
+
+    @Test(groups = { "fast", "emulator" }, timeOut = TIMEOUT)
+    public void UserAgentSuffixWithoutSpecialCharacter() {
+        CosmosClient clientWithUserAgentSuffix = getClientBuilder()
+            .userAgentSuffix("TestUserAgent")
+            .buildClient();
+
+        CosmosContainerResponse response =
+            clientWithUserAgentSuffix.getDatabase(this.databaseName).getContainer(this.containerName).read();
+
+        assertThat(response).isNotNull();
+        assertThat(response.getStatusCode()).isEqualTo(200);
+        assertThat(response.getProperties()).isNotNull();
+        assertThat(response.getProperties().getId()).isEqualTo(this.containerName);
+    }
+
+    @Test(groups = { "fast", "emulator" }, timeOut = TIMEOUT)
+    public void UserAgentSuffixWithSpecialCharacter() {
```

Comment:
Might be worth some other tests too - something that is 4-byte encoded, or that requires a surrogate pair.

Also, something that encodes differently in latin-1 and UTF-8?

Reply:
Added two more test cases - but given that the main algorithm to replace is based on standard Java - normalize + regex replacing any non-ASCII char with"_" it should be fine. This algorithm certainly avoids invalid chars - it does not make it possible to decode back to the original string provided by the user - but instead optimizes for "best" human readability by applying the Normalization. I think that is the right trade-off since we don't have to provide decoding. Alternatives would be UrlEncoding or Punycode - but both don't make it very human-readable -which is the primary goal on how UserAgent suffix is used today for debugging.